lib/Core/Executor.cpp

```diff
@@ -3552,46 +3574,74 @@ std::string Executor::getAddressInfo(ExecutionState &state,
 }
 
 
-void Executor::terminateState(ExecutionState &state) {
+void Executor::terminateState(ExecutionState &state, StateTerminationType terminationType) {
   if (replayKTest && replayPosition!=replayKTest->numObjects) {
     klee_warning_once(replayKTest,
                       "replay did not consume all objects in test input.");
   }
 
+  // update stats
   interpreterHandler->incPathsExplored();
+  if (processTree)
+    state.ptreeNode->terminationTypeMask |= 1U << static_cast<std::uint32_t>(terminationType);
 
-  std::vector<ExecutionState *>::iterator it =
-      std::find(addedStates.begin(), addedStates.end(), &state);
-  if (it==addedStates.end()) {
-    state.pc = state.prevPC;
+  // exit on termination type
+  if (ExitOnErrorType.isSet(terminationType))
+    haltExecution = true;
 
+  // add state to removedStates for next searcher update
+  auto it = std::find(addedStates.begin(), addedStates.end(), &state);
+  if (it == addedStates.end()) {
+    state.pc = state.prevPC;
     removedStates.push_back(&state);
-  } else {
-    // never reached searcher, just delete immediately
-    std::map< ExecutionState*, std::vector<SeedInfo> >::iterator it3 = 
-      seedMap.find(&state);
-    if (it3 != seedMap.end())
-      seedMap.erase(it3);
-    addedStates.erase(it);
-    processTree->remove(state.ptreeNode);
-    delete &state;
+    return;
   }
+  // ... or just delete immediately if never reached searcher
```

Comment:
Perhaphs not a part of this, but I think it would be much better if this optimisation were gone. It would make the code easier to reason about if everything always goes through the searcher.